cmd/test_load.go

```diff
@@ -0,0 +1,199 @@
+package cmd
+
+import (
+	"archive/tar"
+	"bytes"
+	"fmt"
+
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"go.k6.io/k6/errext"
+	"go.k6.io/k6/errext/exitcodes"
+	"go.k6.io/k6/js"
+	"go.k6.io/k6/lib"
+	"go.k6.io/k6/lib/metrics"
+	"go.k6.io/k6/loader"
+)
+
+const (
+	testTypeJS      = "js"
+	testTypeArchive = "archive"
+)
+
+type loadedTest struct {
+	testPath        string // contains the raw string the user supplied
+	source          *loader.SourceData
+	fileSystems     map[string]afero.Fs
+	runtimeOptions  lib.RuntimeOptions
+	metricsRegistry *metrics.Registry
+	builtInMetrics  *metrics.BuiltinMetrics
+	initRunner      lib.Runner // TODO: rename to something more appropriate
+
+	// Only set if cliConfigGetter is supplied to loadTest() or if
+	// consolidateDeriveAndValidateConfig() is manually called.
+	consolidatedConfig Config
+	derivedConfig      Config
+}
+
+func loadTest(
+	gs *globalState, cmd *cobra.Command, args []string,
+	// supply this if you want the test config consolidated and validated
+	cliConfigGetter func(flags *pflag.FlagSet) (Config, error), // TODO: obviate
+) (*loadedTest, error) {
+	if len(args) < 1 {
+		return nil, fmt.Errorf("k6 needs at least one argument to load the test")
+	}
+
+	testPath := args[0]
+	gs.logger.Debugf("Resolving and reading test '%s'...", testPath)
+	src, fileSystems, err := readSource(gs, testPath)
+	if err != nil {
+		return nil, err
+	}
+	resolvedPath := src.URL.String()
+	gs.logger.Debugf("'%s' resolved to '%s' and successfully loaded %d bytes!", testPath, resolvedPath, len(src.Data))
+
+	gs.logger.Debugf("Gathering k6 runtime options...")
+	runtimeOptions, err := getRuntimeOptions(cmd.Flags(), gs.envVars)
+	if err != nil {
+		return nil, err
+	}
+
+	registry := metrics.NewRegistry()
+	test := &loadedTest{
+		testPath:        testPath,
+		source:          src,
+		fileSystems:     fileSystems,
+		runtimeOptions:  runtimeOptions,
+		metricsRegistry: registry,
+		builtInMetrics:  metrics.RegisterBuiltinMetrics(registry),
+	}
+
+	gs.logger.Debugf("Initializing k6 runner for '%s' (%s)...", testPath, resolvedPath)
+	if err := test.initializeFirstRunner(gs); err != nil {
+		return nil, fmt.Errorf("could not initialize '%s': %w", testPath, err)
+	}
+	gs.logger.Debug("Runner successfully initialized!")
+
+	if cliConfigGetter != nil {
+		if err := test.consolidateDeriveAndValidateConfig(gs, cmd, cliConfigGetter); err != nil {
+			return nil, err
+		}
+	}
+
+	return test, nil
+}
+
+func (lt *loadedTest) initializeFirstRunner(gs *globalState) error {
```

Comment:
The name of this function led to me to an immediate load of questions: "why do we load a **first** runner", "how is the **first** runner different from the other runners**. Which leads me to suggest to detail that in the documentation of the function 😸

Reply:
Hmm yeah, though the bigger problem is that `Runner` shouldn't even be called `Runner` :sweat_smile: It's a terrible name and something we've wanted to rename for a long, **long** time, see https://github.com/grafana/k6/blame/3db46be4b1f4591a3ae85b3d4fd863bb94e37165/lib/runner.go#L38

A VUFactory will be a better name, but still quite terrible, since a `Runner` does other things besides creating VUs :sob: So it first actually needs a complete refactor/rewrite, probably to split apart its responsibilities in 2 or 3 different clear and well-defined concepts and interfaces, which we can then name nicely... That is a task for another time though, refactoring the similarly unclear `Engine` was enough for this series of PRs :sweat_smile: Until that future time, I chose to keep the current confusing nomenclature about `Runner` and just sprinkle some TODOs https://github.com/grafana/k6/blob/207ef83e62b7ac097b6f024e9ca5a8ff56d93ba0/cmd/test_load.go#L31

I will add a comment to this method though, to explain what is happening in it and why :+1: